pipeline/relation_embedder/relation_embedder/src/main/scala/uk/ac/wellcome/relation_embedder/ArchiveRelationsCache.scala

```diff
@@ -33,17 +33,47 @@ class ArchiveRelationsCache(works: Map[String, RelationWork]) extends Logging {
         Relations.none
       }
 
+  // The availabilities of an archive's Relations are the union of
+  // all of its descendants' availabilities, as well as its own
+  def getAvailabilities(work: Work[Merged]): Set[Availability] =
+    work.data.collectionPath
+      .map {
+        case CollectionPath(path, _) =>
+          @tailrec
+          def availabilities(stack: List[String],
+                             accum: Set[Availability]): Set[Availability] =
+            stack match {
+              case Nil => accum
+              case head :: tail =>
+                val children = childMapping.getOrElse(head, Nil)
+
+                val childAvailabilities =
+                  (head :: children)
+                    .map(works)
+                    .map(_.state.availabilities)
+                    .foldLeft(Set.empty[Availability])(_ union _)
+
+                availabilities(
+                  childMapping.getOrElse(head, Nil) ++ tail,
+                  accum union childAvailabilities
+                )
+            }
+          availabilities(
+            childMapping.getOrElse(path, Nil),
+            works.get(path).map(_.state.availabilities).getOrElse(Set.empty)
+          )
+      }
+      .getOrElse(Set.empty)
```

Comment:
I'm not quite au fait enough with Scala's FP concepts to be sure this is correct without stepping through it in an IDE, but it looks sensible and I'll give you the benefit of the doubt. 👍